// konsier/dist/adapters/express.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/adapters/express.ts
var express_exports = {};
__export(express_exports, {
  serveKonsier: () => serveKonsier,
  verifyKonsierPage: () => verifyKonsierPage
});
module.exports = __toCommonJS(express_exports);

// src/constants.ts
var DEFAULT_ALLOWED_CLOCK_SKEW_MS = 5 * 60 * 1e3;
var DEFAULT_PAGE_SESSION_TTL_MS = 15 * 60 * 1e3;

// src/protocol/signatures.ts
var import_node_crypto = require("crypto");

// src/tool.ts
var import_zod = require("zod");

// src/client.ts
function createJsonBodyMiddleware(rawBodyProperty) {
  return function jsonBodyMiddleware(req, _res, next) {
    if (typeof next !== "function") {
      return;
    }
    if (typeof req.body !== "undefined") {
      if (typeof req[rawBodyProperty] === "undefined") {
        if (Buffer.isBuffer(req.body)) {
          req[rawBodyProperty] = req.body;
        } else if (typeof req.body === "string") {
          req[rawBodyProperty] = Buffer.from(req.body, "utf8");
        } else if (req.body && typeof req.body === "object") {
          req[rawBodyProperty] = Buffer.from(JSON.stringify(req.body), "utf8");
        }
      }
      next();
      return;
    }
    const method = (req.method ?? "POST").toUpperCase();
    if (method !== "POST") {
      next();
      return;
    }
    if (typeof req.on !== "function") {
      next();
      return;
    }
    const chunks = [];
    let finished = false;
    const finalize = (error) => {
      if (finished) {
        return;
      }
      finished = true;
      next(error);
    };
    req.on("data", (chunk) => {
      chunks.push(
        Buffer.isBuffer(chunk) ? chunk : Buffer.from(typeof chunk === "string" ? chunk : "")
      );
    });
    req.on("end", () => {
      const rawBody = Buffer.concat(chunks);
      req[rawBodyProperty] = rawBody;
      if (rawBody.length === 0) {
        req.body = {};
        finalize();
        return;
      }
      try {
        req.body = JSON.parse(rawBody.toString("utf8"));
      } catch {
        req.body = rawBody.toString("utf8");
      }
      finalize();
    });
    req.on("error", (error) => {
      finalize(error);
    });
  };
}

// src/adapters/express.ts
function serveKonsier(app, konsier) {
  app.post(
    konsier.webhookPath(),
    createJsonBodyMiddleware("rawBody"),
    konsier.webhookHandler()
  );
}
function verifyKonsierPage(konsier) {
  return function pageAuth(req, res, next) {
    const result = konsier.pageRequest({
      url: resolveExpressRequestUrl(req),
      headers: req.headers
    });
    if (result.type === "authorized") {
      req.konsier = result.context;
      next?.();
      return;
    }
    if (typeof res.status === "function") {
      res.status(result.status);
    } else {
      res.statusCode = result.status;
    }
    for (const [name, value] of Object.entries(result.headers)) {
      if (typeof res.setHeader === "function") {
        res.setHeader(name, value);
      }
    }
    if (typeof res.send === "function") {
      res.send(result.body ?? "");
      return;
    }
    res.end?.(result.body ?? "");
  };
}
function resolveExpressRequestUrl(req) {
  const forwardedProto = readFirstHeaderValue(req.headers?.["x-forwarded-proto"]);
  const forwardedHost = readFirstHeaderValue(req.headers?.["x-forwarded-host"]);
  const protocol = forwardedProto || req.protocol || "http";
  const host = forwardedHost || readFirstHeaderValue(req.headers?.host) || "localhost";
  const path = req.originalUrl || req.url || "/";
  return `${protocol}://${host}${path}`;
}
function readFirstHeaderValue(value) {
  if (Array.isArray(value)) {
    return value[0];
  }
  return value;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  serveKonsier,
  verifyKonsierPage
});
