// konsier/dist/adapters/fastify.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/adapters/fastify.ts
var fastify_exports = {};
__export(fastify_exports, {
  registerKonsier: () => registerKonsier,
  verifyKonsierPageRequest: () => verifyKonsierPageRequest
});
module.exports = __toCommonJS(fastify_exports);

// src/adapters/shared.ts
function headersToObject(headers) {
  const normalized = {};
  if (typeof Headers !== "undefined" && headers instanceof Headers) {
    headers.forEach((value, key) => {
      normalized[key] = value;
    });
    return normalized;
  }
  for (const [key, value] of Object.entries(headers)) {
    normalized[key] = value;
  }
  return normalized;
}
function verifyPageRequest(konsier, input) {
  if (input instanceof Request) {
    return konsier.pageRequest({
      url: input.url,
      headers: headersToObject(input.headers)
    });
  }
  return konsier.pageRequest({
    url: input.url,
    headers: headersToObject(input.headers)
  });
}

// src/adapters/fastify.ts
function registerKonsier(fastify, konsier) {
  fastify.route({
    method: "POST",
    url: konsier.webhookPath(),
    config: {
      rawBody: true
    },
    handler: async (request, reply) => {
      await konsier.webhookHandler()(
        {
          method: request.raw?.method ?? request.method,
          headers: request.raw?.headers ?? request.headers,
          body: request.body,
          rawBody: request.rawBody
        },
        createReplyAdapter(reply)
      );
    }
  });
}
function verifyKonsierPageRequest(konsier, request) {
  if (request instanceof Request) {
    return verifyPageRequest(konsier, request);
  }
  if ("url" in request && typeof request.url === "string" && "headers" in request) {
    return verifyPageRequest(konsier, {
      url: request.url,
      headers: request.headers
    });
  }
  const rawRequest = "raw" in request ? request.raw : void 0;
  const headers = rawRequest?.headers ?? request.headers;
  const path = rawRequest?.url ?? request.url ?? "/";
  const protocol = firstHeaderValue(headers["x-forwarded-proto"]) || ("protocol" in request ? request.protocol : void 0) || "http";
  const host = firstHeaderValue(headers["x-forwarded-host"]) || firstHeaderValue(headers.host) || "localhost";
  return verifyPageRequest(konsier, {
    url: `${protocol}://${host}${path}`,
    headers
  });
}
function createReplyAdapter(reply) {
  return {
    status(statusCode) {
      if (typeof reply.status === "function") {
        reply.status(statusCode);
      } else if (typeof reply.code === "function") {
        reply.code(statusCode);
      }
      return this;
    },
    json(body) {
      if (typeof reply.header === "function") {
        reply.header("content-type", "application/json");
      }
      reply.send(body);
      return this;
    },
    send(body) {
      reply.send(body);
      return this;
    },
    end(body) {
      reply.send(body);
      return this;
    },
    setHeader(name, value) {
      if (typeof reply.header === "function") {
        reply.header(name, value);
      }
      return this;
    }
  };
}
function firstHeaderValue(value) {
  if (Array.isArray(value)) {
    return value[0];
  }
  return value;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  registerKonsier,
  verifyKonsierPageRequest
});
