// konsier/dist/adapters/fastify.js
// src/adapters/shared.ts
function headersToObject(headers) {
  const normalized = {};
  if (typeof Headers !== "undefined" && headers instanceof Headers) {
    headers.forEach((value, key) => {
      normalized[key] = value;
    });
    return normalized;
  }
  for (const [key, value] of Object.entries(headers)) {
    normalized[key] = value;
  }
  return normalized;
}
function verifyPageRequest(konsier, input) {
  if (input instanceof Request) {
    return konsier.pageRequest({
      url: input.url,
      headers: headersToObject(input.headers)
    });
  }
  return konsier.pageRequest({
    url: input.url,
    headers: headersToObject(input.headers)
  });
}

// src/adapters/fastify.ts
function registerKonsier(fastify, konsier) {
  fastify.route({
    method: "POST",
    url: konsier.webhookPath(),
    config: {
      rawBody: true
    },
    handler: async (request, reply) => {
      await konsier.webhookHandler()(
        {
          method: request.raw?.method ?? request.method,
          headers: request.raw?.headers ?? request.headers,
          body: request.body,
          rawBody: request.rawBody
        },
        createReplyAdapter(reply)
      );
    }
  });
}
function verifyKonsierPageRequest(konsier, request) {
  if (request instanceof Request) {
    return verifyPageRequest(konsier, request);
  }
  if ("url" in request && typeof request.url === "string" && "headers" in request) {
    return verifyPageRequest(konsier, {
      url: request.url,
      headers: request.headers
    });
  }
  const rawRequest = "raw" in request ? request.raw : void 0;
  const headers = rawRequest?.headers ?? request.headers;
  const path = rawRequest?.url ?? request.url ?? "/";
  const protocol = firstHeaderValue(headers["x-forwarded-proto"]) || ("protocol" in request ? request.protocol : void 0) || "http";
  const host = firstHeaderValue(headers["x-forwarded-host"]) || firstHeaderValue(headers.host) || "localhost";
  return verifyPageRequest(konsier, {
    url: `${protocol}://${host}${path}`,
    headers
  });
}
function createReplyAdapter(reply) {
  return {
    status(statusCode) {
      if (typeof reply.status === "function") {
        reply.status(statusCode);
      } else if (typeof reply.code === "function") {
        reply.code(statusCode);
      }
      return this;
    },
    json(body) {
      if (typeof reply.header === "function") {
        reply.header("content-type", "application/json");
      }
      reply.send(body);
      return this;
    },
    send(body) {
      reply.send(body);
      return this;
    },
    end(body) {
      reply.send(body);
      return this;
    },
    setHeader(name, value) {
      if (typeof reply.header === "function") {
        reply.header(name, value);
      }
      return this;
    }
  };
}
function firstHeaderValue(value) {
  if (Array.isArray(value)) {
    return value[0];
  }
  return value;
}
export {
  registerKonsier,
  verifyKonsierPageRequest
};
