// konsier/dist/adapters/hono.js
// src/adapters/shared.ts
function headersToObject(headers) {
  const normalized = {};
  if (typeof Headers !== "undefined" && headers instanceof Headers) {
    headers.forEach((value, key) => {
      normalized[key] = value;
    });
    return normalized;
  }
  for (const [key, value] of Object.entries(headers)) {
    normalized[key] = value;
  }
  return normalized;
}
async function handleFetchWebhook(konsier, request) {
  const rawBody = await request.text();
  const recorder = createResponseRecorder();
  await konsier.webhookHandler()(
    {
      method: request.method,
      headers: headersToObject(request.headers),
      body: rawBody,
      rawBody
    },
    recorder
  );
  return recorder.toResponse();
}
function verifyPageRequest(konsier, input) {
  if (input instanceof Request) {
    return konsier.pageRequest({
      url: input.url,
      headers: headersToObject(input.headers)
    });
  }
  return konsier.pageRequest({
    url: input.url,
    headers: headersToObject(input.headers)
  });
}
function pageResultToResponse(result) {
  return new Response(result.body ?? null, {
    status: result.status,
    headers: result.headers
  });
}
function createResponseRecorder() {
  return {
    statusCode: 200,
    body: null,
    headers: new Headers(),
    status(code) {
      this.statusCode = code;
      return this;
    },
    json(payload) {
      this.headers.set("content-type", "application/json");
      this.body = JSON.stringify(payload);
      return this;
    },
    send(payload) {
      if (typeof payload === "string" || payload instanceof Uint8Array) {
        this.body = payload;
        return this;
      }
      this.headers.set("content-type", "application/json");
      this.body = JSON.stringify(payload);
      return this;
    },
    end(payload) {
      if (typeof payload === "undefined") {
        this.body = null;
        return this;
      }
      if (typeof payload === "string" || payload instanceof Uint8Array) {
        this.body = payload;
        return this;
      }
      this.body = JSON.stringify(payload);
      return this;
    },
    setHeader(name, value) {
      this.headers.set(name, value);
      return this;
    },
    toResponse() {
      return new Response(this.body, {
        status: this.statusCode,
        headers: this.headers
      });
    }
  };
}

// src/adapters/hono.ts
function konsierWebhook(konsier) {
  return async function handleKonsierWebhook(context) {
    const request = context.req instanceof Request ? context.req : context.req.raw;
    return handleFetchWebhook(konsier, request);
  };
}
function serveKonsier(app, konsier) {
  app.post(konsier.webhookPath(), konsierWebhook(konsier));
}
function verifyKonsierPageRequest(konsier, request) {
  const result = verifyPageRequest(konsier, request);
  if (result.type === "response") {
    return pageResultToResponse(result);
  }
  return result;
}
export {
  konsierWebhook,
  serveKonsier,
  verifyKonsierPageRequest
};
