// konsier/dist/adapters/next.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/adapters/next.ts
var next_exports = {};
__export(next_exports, {
  createKonsierRoute: () => createKonsierRoute,
  verifyKonsierPageRequest: () => verifyKonsierPageRequest
});
module.exports = __toCommonJS(next_exports);

// src/adapters/shared.ts
function headersToObject(headers) {
  const normalized = {};
  if (typeof Headers !== "undefined" && headers instanceof Headers) {
    headers.forEach((value, key) => {
      normalized[key] = value;
    });
    return normalized;
  }
  for (const [key, value] of Object.entries(headers)) {
    normalized[key] = value;
  }
  return normalized;
}
async function handleFetchWebhook(konsier, request) {
  const rawBody = await request.text();
  const recorder = createResponseRecorder();
  await konsier.webhookHandler()(
    {
      method: request.method,
      headers: headersToObject(request.headers),
      body: rawBody,
      rawBody
    },
    recorder
  );
  return recorder.toResponse();
}
function verifyPageRequest(konsier, input) {
  if (input instanceof Request) {
    return konsier.pageRequest({
      url: input.url,
      headers: headersToObject(input.headers)
    });
  }
  return konsier.pageRequest({
    url: input.url,
    headers: headersToObject(input.headers)
  });
}
function pageResultToResponse(result) {
  return new Response(result.body ?? null, {
    status: result.status,
    headers: result.headers
  });
}
function createResponseRecorder() {
  return {
    statusCode: 200,
    body: null,
    headers: new Headers(),
    status(code) {
      this.statusCode = code;
      return this;
    },
    json(payload) {
      this.headers.set("content-type", "application/json");
      this.body = JSON.stringify(payload);
      return this;
    },
    send(payload) {
      if (typeof payload === "string" || payload instanceof Uint8Array) {
        this.body = payload;
        return this;
      }
      this.headers.set("content-type", "application/json");
      this.body = JSON.stringify(payload);
      return this;
    },
    end(payload) {
      if (typeof payload === "undefined") {
        this.body = null;
        return this;
      }
      if (typeof payload === "string" || payload instanceof Uint8Array) {
        this.body = payload;
        return this;
      }
      this.body = JSON.stringify(payload);
      return this;
    },
    setHeader(name, value) {
      this.headers.set(name, value);
      return this;
    },
    toResponse() {
      return new Response(this.body, {
        status: this.statusCode,
        headers: this.headers
      });
    }
  };
}

// src/adapters/next.ts
function createKonsierRoute(konsier) {
  return async function POST(request) {
    return handleFetchWebhook(konsier, request);
  };
}
function verifyKonsierPageRequest(konsier, request) {
  const result = verifyPageRequest(konsier, request);
  if (result.type === "response") {
    return pageResultToResponse(result);
  }
  return result;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createKonsierRoute,
  verifyKonsierPageRequest
});
