// konsier/dist/index.js
// src/constants.ts
var DEFAULT_CLOUD_BASE_URL = "https://konsier.com/api";
var DEFAULT_REQUEST_TIMEOUT_MS = 1e4;
var DEFAULT_MAX_RETRIES = 1;
var DEFAULT_ALLOWED_CLOCK_SKEW_MS = 5 * 60 * 1e3;
var DEFAULT_PAGE_SESSION_TTL_MS = 15 * 60 * 1e3;
var ENV_CLOUD_BASE_URL = "KONSIER_API_BASE_URL";
var HEADER_AUTHORIZATION = "authorization";
var HEADER_CONTENT_TYPE = "content-type";
var HEADER_SIGNATURE = "x-konsier-signature";
var HEADER_TIMESTAMP = "x-konsier-timestamp";
var SIGNATURE_PREFIX = "sha256=";
var PAGE_LAUNCH_QUERY_PARAM = "__konsier";
var PAGE_SESSION_COOKIE_NAME = "__konsier_page";

// src/errors.ts
var KonsierError = class extends Error {
  code;
  statusCode;
  details;
  constructor(input) {
    super(input.message, input.cause ? { cause: input.cause } : void 0);
    this.name = "KonsierError";
    this.code = input.code;
    this.statusCode = input.statusCode ?? 500;
    this.details = input.details;
  }
};
function asKonsierError(error) {
  if (error instanceof KonsierError) {
    return error;
  }
  if (error instanceof Error) {
    return new KonsierError({
      code: "INTERNAL_ERROR",
      message: error.message,
      statusCode: 500,
      cause: error
    });
  }
  return new KonsierError({
    code: "INTERNAL_ERROR",
    message: "Unknown error",
    statusCode: 500,
    details: error
  });
}
function toErrorMessage(error) {
  if (error instanceof Error) {
    return error.message;
  }
  if (typeof error === "string") {
    return error;
  }
  return "Unknown error";
}

// src/cloud/link-user.ts
function requireNonEmptyString(value, input) {
  const normalized = value?.trim() ?? "";
  if (!normalized) {
    throw new KonsierError({
      code: input.code,
      message: input.message,
      statusCode: 400
    });
  }
  return normalized;
}
function normalizeSdkUser(raw) {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new KonsierError({
      code: "INVALID_CLOUD_RESPONSE",
      message: "users.get() returned an invalid payload.",
      statusCode: 500,
      details: raw
    });
  }
  const record = raw;
  const id = typeof record.id === "string" ? record.id.trim() : "";
  if (!id) {
    throw new KonsierError({
      code: "INVALID_CLOUD_RESPONSE",
      message: "users.get() returned an invalid payload.",
      statusCode: 500,
      details: raw
    });
  }
  return {
    id,
    externalId: typeof record.externalId === "string" ? record.externalId : null,
    metadata: record.metadata && typeof record.metadata === "object" && !Array.isArray(record.metadata) ? record.metadata : {},
    firstName: typeof record.firstName === "string" ? record.firstName : null,
    lastName: typeof record.lastName === "string" ? record.lastName : null,
    email: typeof record.email === "string" ? record.email : null,
    phoneNumber: typeof record.phoneNumber === "string" ? record.phoneNumber : null,
    displayName: typeof record.displayName === "string" ? record.displayName : null
  };
}
function normalizeSdkAccount(raw) {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new KonsierError({
      code: "INVALID_CLOUD_RESPONSE",
      message: "accounts API returned an invalid payload.",
      statusCode: 500,
      details: raw
    });
  }
  const record = raw;
  const id = typeof record.id === "string" ? record.id.trim() : "";
  const name = typeof record.name === "string" ? record.name.trim() : "";
  if (!id || !name) {
    throw new KonsierError({
      code: "INVALID_CLOUD_RESPONSE",
      message: "accounts API returned an invalid payload.",
      statusCode: 500,
      details: raw
    });
  }
  const internalRecord = record.internal && typeof record.internal === "object" && !Array.isArray(record.internal) ? record.internal : null;
  return {
    id,
    name,
    logoUrl: typeof record.logoUrl === "string" ? record.logoUrl : null,
    externalId: typeof record.externalId === "string" ? record.externalId : null,
    metadata: record.metadata && typeof record.metadata === "object" && !Array.isArray(record.metadata) ? record.metadata : {},
    connectedAt: typeof record.connectedAt === "string" ? record.connectedAt : "",
    linkedAgents: Array.isArray(record.linkedAgents) ? record.linkedAgents.map((item) => {
      if (!item || typeof item !== "object" || Array.isArray(item)) {
        return null;
      }
      const entry = item;
      const ref = typeof entry.ref === "string" ? entry.ref.trim() : "";
      const agentId = typeof entry.agentId === "string" ? entry.agentId.trim() : "";
      const agentName = typeof entry.agentName === "string" ? entry.agentName.trim() : "";
      if (!ref || !agentId || !agentName) {
        return null;
      }
      return { ref, agentId, agentName };
    }).filter(
      (item) => item !== null
    ) : [],
    internal: {
      pages: Array.isArray(internalRecord?.pages) ? internalRecord.pages.filter(
        (value) => typeof value === "string"
      ) : [],
      tools: Array.isArray(internalRecord?.tools) ? internalRecord.tools.filter(
        (value) => typeof value === "string"
      ) : []
    }
  };
}
async function getUser(client, input) {
  const userId = requireNonEmptyString(input.userId, {
    code: "INVALID_USER_INPUT",
    message: "users.get() requires userId."
  });
  const payload = await client.get(`/users/${encodeURIComponent(userId)}`);
  return normalizeSdkUser(payload.user);
}
async function linkUser(client, input) {
  const userId = requireNonEmptyString(input.userId, {
    code: "INVALID_USER_INPUT",
    message: "users.link() requires userId."
  });
  const externalId = requireNonEmptyString(input.externalId, {
    code: "INVALID_USER_INPUT",
    message: "users.link() requires externalId."
  });
  const payload = await client.post("/users/link", {
    userId,
    externalId,
    metadata: input.metadata ?? {}
  });
  return normalizeSdkUser(payload.user);
}
async function listAccounts(client) {
  const payload = await client.get("/accounts");
  const items = Array.isArray(payload.accounts) ? payload.accounts : [];
  return items.map((item) => normalizeSdkAccount(item));
}
async function getAccount(client, input) {
  const accountId = requireNonEmptyString(input.accountId, {
    code: "INVALID_ACCOUNT_INPUT",
    message: "accounts.get() requires accountId."
  });
  const payload = await client.get(`/accounts/${encodeURIComponent(accountId)}`);
  return normalizeSdkAccount(payload.account);
}
async function linkAccount(client, input) {
  const accountId = requireNonEmptyString(input.accountId, {
    code: "INVALID_ACCOUNT_INPUT",
    message: "accounts.link() requires accountId."
  });
  const externalId = requireNonEmptyString(input.externalId, {
    code: "INVALID_ACCOUNT_INPUT",
    message: "accounts.link() requires externalId."
  });
  const payload = await client.post("/accounts/link", {
    accountId,
    externalId,
    metadata: input.metadata ?? {}
  });
  return normalizeSdkAccount(payload.account);
}
async function startConnection(client, input) {
  const redirect = requireNonEmptyString(input.redirect, {
    code: "INVALID_CONNECTION_INPUT",
    message: "connections.start() requires redirect."
  });
  const payload = await client.post("/connections/start", {
    redirect,
    metadata: input.metadata ?? {}
  });
  const url = typeof payload.url === "string" ? payload.url : "";
  const expiresAt = typeof payload.expiresAt === "string" ? payload.expiresAt : "";
  if (!url || !expiresAt) {
    throw new KonsierError({
      code: "INVALID_CLOUD_RESPONSE",
      message: "connections.start() returned an invalid payload.",
      statusCode: 500,
      details: payload
    });
  }
  return { url, expiresAt };
}
async function completeConnection(client, input) {
  const token = requireNonEmptyString(input.token, {
    code: "INVALID_CONNECTION_INPUT",
    message: "connections.complete() requires token."
  });
  const payload = await client.post("/connections/complete", {
    token
  });
  return {
    account: normalizeSdkAccount(payload.account)
  };
}

// src/cloud/http.ts
var CloudApiClient = class {
  apiKey;
  baseUrl;
  timeoutMs;
  maxRetries;
  debug;
  constructor(options) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.timeoutMs = DEFAULT_REQUEST_TIMEOUT_MS;
    this.maxRetries = DEFAULT_MAX_RETRIES;
    this.debug = Boolean(options.debug);
  }
  async get(path) {
    return this.request("GET", path);
  }
  async post(path, body) {
    return this.request("POST", path, body);
  }
  async request(method, path, body) {
    const url = `${this.baseUrl}${path.startsWith("/") ? path : `/${path}`}`;
    const normalizedBody = typeof body === "undefined" ? void 0 : stripUndefined(body);
    const payload = typeof normalizedBody === "undefined" ? void 0 : JSON.stringify(normalizedBody);
    if (shouldDebugLog(this.debug)) {
      console.log("[konsier] cloud request", {
        url,
        method,
        path,
        body: normalizedBody
      });
    }
    let lastError = null;
    for (let attempt = 0; attempt <= this.maxRetries; attempt += 1) {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
      try {
        const requestInit = {
          method,
          headers: {
            [HEADER_AUTHORIZATION]: `Bearer ${this.apiKey}`,
            ...payload ? { [HEADER_CONTENT_TYPE]: "application/json" } : {}
          },
          signal: controller.signal
        };
        if (payload) {
          requestInit.body = payload;
        }
        const response = await fetch(url, requestInit);
        const raw = await response.text();
        const parsed = parseJsonObject(raw);
        if (!response.ok) {
          if (response.status >= 500 && attempt < this.maxRetries) {
            continue;
          }
          const message = typeof parsed?.error === "string" ? parsed.error : raw || `Cloud request failed (${response.status})`;
          throw new KonsierError({
            code: "CLOUD_REQUEST_FAILED",
            message,
            statusCode: response.status,
            details: parsed
          });
        }
        if (shouldDebugLog(this.debug)) {
          console.log("[konsier] cloud response", {
            url,
            status: response.status,
            body: parsed ?? {}
          });
        }
        return parsed ?? {};
      } catch (error) {
        lastError = error;
        const shouldRetry = attempt < this.maxRetries && (error instanceof TypeError || error instanceof Error && error.name === "AbortError");
        if (!shouldRetry) {
          if (shouldDebugLog(this.debug)) {
            console.warn("[konsier] cloud request failed", {
              url,
              attempt: attempt + 1,
              error: error instanceof Error ? error.message : "Unknown error"
            });
          }
          throw error;
        }
      } finally {
        clearTimeout(timeout);
      }
    }
    throw lastError ?? new KonsierError({
      code: "CLOUD_REQUEST_FAILED",
      message: "Cloud request failed"
    });
  }
};
function resolveCloudBaseUrl(input) {
  const override = process.env[ENV_CLOUD_BASE_URL]?.trim();
  if (override) {
    const resolved = override.replace(/\/+$/, "");
    if (shouldDebugLog(Boolean(input?.debug))) {
      console.log("[konsier] resolved cloud base URL from env", {
        env: ENV_CLOUD_BASE_URL,
        value: resolved
      });
    }
    return resolved;
  }
  if (shouldDebugLog(Boolean(input?.debug))) {
    console.log("[konsier] resolved cloud base URL from default", {
      value: DEFAULT_CLOUD_BASE_URL
    });
  }
  return DEFAULT_CLOUD_BASE_URL;
}
function shouldDebugLog(debug) {
  return debug && process.env.NODE_ENV === "development";
}
function parseJsonObject(raw) {
  if (!raw) {
    return {};
  }
  try {
    const parsed = JSON.parse(raw);
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
      return parsed;
    }
    return { value: parsed };
  } catch {
    return { raw };
  }
}
function stripUndefined(value) {
  const out = {};
  for (const [key, entry] of Object.entries(value)) {
    if (entry !== void 0) {
      out[key] = entry;
    }
  }
  return out;
}

// src/cloud/send.ts
async function sendMessage(client, input) {
  if (!input.userId && !input.conversationId) {
    throw new KonsierError({
      code: "INVALID_SEND_INPUT",
      message: "send() requires either userId or conversationId.",
      statusCode: 400
    });
  }
  if (!input.text && !input.html && (!input.attachments || input.attachments.length === 0)) {
    throw new KonsierError({
      code: "INVALID_SEND_INPUT",
      message: "send() requires text, html, or at least one attachment.",
      statusCode: 400
    });
  }
  await client.post("/messages/send", {
    userId: input.userId,
    conversationId: input.conversationId === void 0 ? void 0 : String(input.conversationId),
    text: input.text,
    html: input.html,
    attachments: input.attachments
  });
}

// src/protocol/signatures.ts
import { createHmac, timingSafeEqual } from "crypto";
function getHeaderValue(headers, name) {
  const targetName = name.toLowerCase();
  for (const [headerName, headerValue] of Object.entries(headers)) {
    if (headerName.toLowerCase() !== targetName) {
      continue;
    }
    if (Array.isArray(headerValue)) {
      const firstValue = headerValue[0];
      return typeof firstValue === "string" ? firstValue : void 0;
    }
    return headerValue;
  }
  return void 0;
}
function normalizeSignature(signature) {
  const trimmed = signature.trim();
  if (trimmed.toLowerCase().startsWith(SIGNATURE_PREFIX)) {
    return trimmed.slice(SIGNATURE_PREFIX.length);
  }
  return trimmed;
}
function createKonsierSignature(input) {
  return createHmac("sha256", input.apiKey).update(`${input.timestamp}.${input.payload}`).digest("hex");
}
function verifyKonsierSignature(input) {
  const timestampMs = Number.parseInt(input.timestamp, 10);
  if (!Number.isFinite(timestampMs)) {
    return { ok: false, reason: "INVALID_TIMESTAMP" };
  }
  const now = input.nowMs ?? Date.now();
  const skew = input.allowedClockSkewMs ?? DEFAULT_ALLOWED_CLOCK_SKEW_MS;
  if (Math.abs(now - timestampMs) > skew) {
    return { ok: false, reason: "TIMESTAMP_OUT_OF_RANGE" };
  }
  const expected = createKonsierSignature({
    apiKey: input.apiKey,
    timestamp: input.timestamp,
    payload: input.payload
  });
  const provided = normalizeSignature(input.providedSignature);
  try {
    const expectedBuffer = Buffer.from(expected, "hex");
    const providedBuffer = Buffer.from(provided, "hex");
    if (expectedBuffer.length !== providedBuffer.length) {
      return { ok: false, reason: "SIGNATURE_MISMATCH" };
    }
    if (!timingSafeEqual(expectedBuffer, providedBuffer)) {
      return { ok: false, reason: "SIGNATURE_MISMATCH" };
    }
    return { ok: true };
  } catch {
    return { ok: false, reason: "SIGNATURE_PARSE_ERROR" };
  }
}
function createPageSessionToken(input) {
  return createSignedPageToken(input.apiKey, {
    type: "session",
    projectId: input.projectId ?? null,
    account: normalizePageTokenAccount(input.account),
    user: normalizePageTokenUser(input.user),
    theme: normalizePageTheme(input.theme),
    exp: input.exp
  });
}
function verifyPageToken(input) {
  const dotIndex = input.token.indexOf(".");
  if (dotIndex <= 0 || dotIndex === input.token.length - 1) {
    return { ok: false, reason: "INVALID_TOKEN_FORMAT" };
  }
  const encodedPayload = input.token.slice(0, dotIndex);
  const providedSignature = input.token.slice(dotIndex + 1);
  const expectedSignature = createHmac("sha256", input.apiKey).update(encodedPayload).digest("hex");
  try {
    const expectedBuffer = Buffer.from(expectedSignature, "hex");
    const providedBuffer = Buffer.from(providedSignature, "hex");
    if (expectedBuffer.length !== providedBuffer.length) {
      return { ok: false, reason: "TOKEN_SIGNATURE_MISMATCH" };
    }
    if (!timingSafeEqual(expectedBuffer, providedBuffer)) {
      return { ok: false, reason: "TOKEN_SIGNATURE_MISMATCH" };
    }
  } catch {
    return { ok: false, reason: "TOKEN_SIGNATURE_PARSE_ERROR" };
  }
  try {
    const decoded = Buffer.from(encodedPayload, "base64url").toString("utf8");
    const parsed = JSON.parse(decoded);
    if (!parsed || typeof parsed !== "object") {
      return { ok: false, reason: "INVALID_TOKEN_PAYLOAD" };
    }
    if (typeof parsed.exp !== "number" || !Number.isFinite(parsed.exp) || parsed.exp <= (input.nowMs ?? Date.now())) {
      return { ok: false, reason: "TOKEN_EXPIRED" };
    }
    return { ok: true, payload: parsed };
  } catch {
    return { ok: false, reason: "INVALID_TOKEN_PAYLOAD" };
  }
}
function createSignedPageToken(apiKey, payload) {
  const encodedPayload = Buffer.from(
    JSON.stringify(payload),
    "utf8"
  ).toString("base64url");
  const signature = createHmac("sha256", apiKey).update(encodedPayload).digest("hex");
  return `${encodedPayload}.${signature}`;
}
function normalizePageTokenAccount(input) {
  if (!input) {
    return null;
  }
  return {
    id: input.id ?? null,
    name: input.name ?? null,
    metadata: input.metadata ?? {}
  };
}
function normalizePageTokenUser(input) {
  return {
    id: input?.id ?? null,
    email: input?.email ?? null,
    name: input?.name ?? null
  };
}
function normalizePageTheme(input) {
  return input === "dark" ? "dark" : "light";
}

// src/handler/resolve-agent.ts
function asResolveAgentRequest(payload) {
  const obj = asObject(payload);
  if (!obj || obj.type !== "resolve_agent") {
    return null;
  }
  if (typeof obj.agent !== "string" || obj.agent.trim().length === 0) {
    return null;
  }
  const account = parseInboundAccount(obj.account);
  if (obj.account !== null && obj.account !== void 0 && account === null) {
    return null;
  }
  return {
    type: "resolve_agent",
    agent: obj.agent,
    account
  };
}
async function resolveAgentRequest(request, dependencies) {
  const resolved = await dependencies.resolveAgentConfig(
    request.agent,
    normalizeAccount(request.account)
  );
  if (!resolved.systemPrompt?.trim()) {
    throw new KonsierError({
      code: "INVALID_AGENT_CONFIG",
      message: `Agent "${request.agent}" has an empty systemPrompt.`,
      statusCode: 500
    });
  }
  return {
    systemPrompt: resolved.systemPrompt,
    tools: resolved.tools.map((tool) => ({
      name: tool.name,
      description: tool.description,
      input: tool.inputSchema
    }))
  };
}
function normalizeAccount(account) {
  if (!account) {
    return null;
  }
  return {
    id: String(account.id),
    name: account.name,
    metadata: account.metadata ?? {}
  };
}
function asObject(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  return value;
}
function parseInboundAccount(value) {
  if (value === null || value === void 0) {
    return null;
  }
  const obj = asObject(value);
  if (!obj) {
    return null;
  }
  const id = obj.id;
  const name = obj.name;
  const metadata = obj.metadata;
  if (!(typeof id === "string" || typeof id === "number")) {
    return null;
  }
  if (typeof name !== "string") {
    return null;
  }
  if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
    return null;
  }
  return {
    id,
    name,
    metadata
  };
}

// src/handler/tool-call.ts
var CHANNELS = /* @__PURE__ */ new Set([
  "telegram",
  "slack",
  "discord",
  "whatsapp",
  "email",
  "sms",
  "konsier"
]);
function asToolCallRequest(payload) {
  const obj = asObject2(payload);
  if (!obj || obj.type !== "tool_call") {
    return null;
  }
  const target = parseTarget(obj.target);
  const channel = parseChannel(obj.channel);
  const conversation = parseConversation(obj.conversation);
  const message = parseMessage(obj.message);
  const tool = parseTool(obj.tool);
  const account = parseInboundAccount2(obj.account);
  const user = parseInboundUser(obj.user);
  if (!target || !channel || !conversation || !message || !tool) {
    return null;
  }
  if (obj.account !== null && obj.account !== void 0 && account === null) {
    return null;
  }
  if (obj.user !== null && obj.user !== void 0 && user === null) {
    return null;
  }
  return {
    type: "tool_call",
    conversation,
    message,
    channel,
    target,
    tool,
    account,
    user
  };
}
async function executeToolCallRequest(request, dependencies) {
  const debug = Boolean(dependencies.debug);
  const account = normalizeAccount2(request.account);
  const resolvedTool = await resolveTool(request, dependencies, account);
  const parsedInput = resolvedTool.parseInput(request.tool.input);
  const user = normalizeUser(request.user);
  debugLog(debug, "tool execution started", {
    target: request.target,
    tool: request.tool.name,
    input: parsedInput,
    conversationId: request.conversation.id,
    channel: request.channel,
    account,
    user
  });
  const context = {
    channel: request.channel,
    agent: request.target.type === "agent" ? request.target.agent : "internal",
    user,
    conversation: {
      id: String(request.conversation.id),
      startedAt: request.conversation.started_at,
      messageCount: request.conversation.message_count
    },
    message: request.message,
    account,
    send: async (message) => {
      debugLog(debug, "tool context.send invoked", {
        target: request.target,
        tool: request.tool.name,
        conversationId: request.conversation.id,
        userId: user.id,
        message
      });
      await dependencies.sendMessage({
        conversationId: request.conversation.id,
        userId: user.id,
        ...message
      });
    }
  };
  try {
    const result = await resolvedTool.handler(parsedInput, context);
    const output = assertToolOutput(result);
    debugLog(debug, "tool execution succeeded", {
      target: request.target,
      tool: request.tool.name,
      input: parsedInput,
      output,
      conversationId: request.conversation.id,
      channel: request.channel
    });
    return output;
  } catch (error) {
    debugLog(debug, "tool execution failed", {
      target: request.target,
      tool: request.tool.name,
      input: parsedInput,
      conversationId: request.conversation.id,
      channel: request.channel,
      error: error instanceof Error ? error.message : "Unknown error",
      errorDetails: error
    });
    throw error;
  }
}
async function resolveTool(request, dependencies, account) {
  if (request.target.type === "agent") {
    const agent = await dependencies.resolveAgentConfig(request.target.agent, account);
    const tool2 = agent.tools.find((entry) => entry.name === request.tool.name);
    if (!tool2) {
      throw new KonsierError({
        code: "TOOL_NOT_FOUND",
        message: `Tool "${request.tool.name}" is not registered for agent "${request.target.agent}".`,
        statusCode: 404
      });
    }
    return tool2;
  }
  const internal = await dependencies.resolveInternalDefinition(account);
  const tool = (internal.tools ?? []).find((entry) => entry.name === request.tool.name);
  if (!tool) {
    throw new KonsierError({
      code: "TOOL_NOT_FOUND",
      message: `Internal tool "${request.tool.name}" is not registered.`,
      statusCode: 404
    });
  }
  return tool;
}
function assertToolOutput(output) {
  if (output && typeof output === "object" && !Array.isArray(output)) {
    return output;
  }
  throw new KonsierError({
    code: "INVALID_TOOL_OUTPUT",
    message: "Tool handlers must return a JSON object.",
    statusCode: 500
  });
}
function normalizeAccount2(account) {
  if (!account) {
    return null;
  }
  return {
    id: String(account.id),
    name: account.name,
    metadata: account.metadata ?? {}
  };
}
function normalizeUser(user) {
  if (!user) {
    return {
      id: "unknown"
    };
  }
  const normalized = {
    id: String(user.id)
  };
  if (user.external_id !== null) {
    normalized.externalId = user.external_id;
  }
  if (user.metadata !== null) {
    normalized.metadata = user.metadata;
  }
  if (user.name !== null) {
    normalized.displayName = user.name;
  }
  return normalized;
}
function debugLog(debug, message, meta) {
  if (!debug || process.env.NODE_ENV !== "development") {
    return;
  }
  console.log("[konsier] tool_call", meta ? { message, ...meta } : { message });
}
function asObject2(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  return value;
}
function parseTarget(value) {
  const obj = asObject2(value);
  if (!obj || typeof obj.type !== "string") {
    return null;
  }
  if (obj.type === "internal") {
    return { type: "internal" };
  }
  if (obj.type === "agent" && typeof obj.agent === "string" && obj.agent.trim()) {
    return {
      type: "agent",
      agent: obj.agent.trim()
    };
  }
  return null;
}
function parseChannel(value) {
  if (typeof value !== "string" || !CHANNELS.has(value)) {
    return null;
  }
  return value;
}
function parseConversation(value) {
  const obj = asObject2(value);
  if (!obj) {
    return null;
  }
  const id = obj.id;
  const projectId = obj.project_id;
  const executionProjectId = obj.execution_project_id;
  const startedAt = obj.started_at;
  const messageCount = obj.message_count;
  if (typeof id !== "number" || typeof projectId !== "number" || typeof executionProjectId !== "number" || typeof startedAt !== "string" || typeof messageCount !== "number") {
    return null;
  }
  return {
    id,
    project_id: projectId,
    execution_project_id: executionProjectId,
    started_at: startedAt,
    message_count: messageCount
  };
}
function parseMessage(value) {
  if (value === null || value === void 0) {
    return {};
  }
  const obj = asObject2(value);
  if (!obj) {
    return null;
  }
  const text = obj.text;
  const html = obj.html;
  const attachments = obj.attachments;
  if (!(typeof text === "undefined" || typeof text === "string") || !(typeof html === "undefined" || typeof html === "string")) {
    return null;
  }
  const normalized = {};
  if (typeof text === "string") {
    normalized.text = text;
  }
  if (typeof html === "string") {
    normalized.html = html;
  }
  if (typeof attachments !== "undefined") {
    const parsedAttachments = parseAttachments(attachments);
    if (!parsedAttachments) {
      return null;
    }
    normalized.attachments = parsedAttachments;
  }
  return normalized;
}
function parseAttachments(value) {
  if (!Array.isArray(value)) {
    return null;
  }
  const attachments = [];
  for (const item of value) {
    const record = asObject2(item);
    if (!record) {
      return null;
    }
    if (typeof record.url !== "string" || !["image", "file", "video", "audio"].includes(String(record.type))) {
      return null;
    }
    const attachment = {
      url: record.url,
      type: record.type
    };
    if (typeof record.name === "string") {
      attachment.name = record.name;
    }
    if (typeof record.mimeType === "string") {
      attachment.mimeType = record.mimeType;
    }
    attachments.push(attachment);
  }
  return attachments;
}
function parseTool(value) {
  const obj = asObject2(value);
  if (!obj) {
    return null;
  }
  const name = obj.name;
  const input = obj.input;
  if (typeof name !== "string") {
    return null;
  }
  if (!input || typeof input !== "object" || Array.isArray(input)) {
    return null;
  }
  return {
    name,
    input
  };
}
function parseInboundAccount2(value) {
  if (value === null || value === void 0) {
    return null;
  }
  const obj = asObject2(value);
  if (!obj) {
    return null;
  }
  const id = obj.id;
  const name = obj.name;
  const metadata = obj.metadata;
  if (!(typeof id === "string" || typeof id === "number")) {
    return null;
  }
  if (typeof name !== "string") {
    return null;
  }
  if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
    return null;
  }
  return {
    id,
    name,
    metadata
  };
}
function parseInboundUser(value) {
  if (value === null || value === void 0) {
    return null;
  }
  const obj = asObject2(value);
  if (!obj) {
    return null;
  }
  const id = obj.id;
  const externalId = obj.external_id;
  const name = obj.name;
  const metadata = obj.metadata;
  if (!(typeof id === "string" || typeof id === "number")) {
    return null;
  }
  if (!(externalId === null || typeof externalId === "string")) {
    return null;
  }
  if (!(name === null || typeof name === "string")) {
    return null;
  }
  if (!(metadata === null || typeof metadata === "object" && !Array.isArray(metadata))) {
    return null;
  }
  return {
    id,
    external_id: externalId,
    name,
    metadata
  };
}

// src/handler/index.ts
function createHandler(dependencies) {
  const rawBodyProperty = dependencies.rawBodyProperty ?? "rawBody";
  const allowedClockSkewMs = dependencies.allowedClockSkewMs ?? DEFAULT_ALLOWED_CLOCK_SKEW_MS;
  const debug = Boolean(dependencies.debug);
  return async function konsierHandler(req, res, next) {
    try {
      if ((req.method ?? "POST").toUpperCase() !== "POST") {
        debugLog2(debug, "method rejected", { method: req.method ?? "POST" });
        sendJson(res, 405, { error: "Method not allowed" });
        return;
      }
      const signature = getHeaderValue(req.headers, HEADER_SIGNATURE);
      const timestamp = getHeaderValue(req.headers, HEADER_TIMESTAMP);
      if (!signature || !timestamp) {
        debugLog2(debug, "missing signature headers");
        sendJson(res, 401, { error: "Missing signature headers" });
        return;
      }
      const rawBody = extractRawBody(req, rawBodyProperty);
      const verified = verifyKonsierSignature({
        apiKey: dependencies.apiKey,
        timestamp,
        payload: rawBody,
        providedSignature: signature,
        allowedClockSkewMs
      });
      if (!verified.ok) {
        debugLog2(debug, "signature verification failed", {
          reason: verified.reason
        });
        sendJson(res, 401, { error: verified.reason });
        return;
      }
      const payload = parseBody(req.body, rawBody);
      if (!payload) {
        debugLog2(debug, "invalid JSON body");
        sendJson(res, 400, { error: "Invalid JSON request body" });
        return;
      }
      debugLog2(debug, "request received", {
        type: payload.type
      });
      const toolCall = asToolCallRequest(payload);
      if (toolCall) {
        const response = await executeToolCallRequest(
          toolCall,
          dependencies.handleToolCall
        );
        debugLog2(debug, "tool_call handled", {
          agent: toolCall.target.type === "agent" ? toolCall.target.agent : null,
          tool: toolCall.tool
        });
        sendJson(res, 200, response);
        return;
      }
      if (payload.type === "tool_call") {
        debugLog2(debug, "invalid tool_call payload");
        sendJson(res, 400, {
          error: "Invalid tool_call request",
          code: "INVALID_TOOL_CALL_REQUEST"
        });
        return;
      }
      const resolveAgent = asResolveAgentRequest(payload);
      if (resolveAgent) {
        const response = await resolveAgentRequest(
          resolveAgent,
          dependencies.handleResolveAgent
        );
        debugLog2(debug, "resolve_agent handled", {
          agent: resolveAgent.agent
        });
        sendJson(res, 200, response);
        return;
      }
      if (payload.type === "resolve_agent") {
        debugLog2(debug, "invalid resolve_agent payload");
        sendJson(res, 400, {
          error: "Invalid resolve_agent request",
          code: "INVALID_RESOLVE_AGENT_REQUEST"
        });
        return;
      }
      const manifest = asManifestRequest(payload);
      if (manifest) {
        const response = await dependencies.handleManifest.listManifest(
          normalizeManifestAccount(manifest.account)
        );
        debugLog2(debug, "manifest handled", {
          accountId: manifest.account?.id ?? null,
          agentCount: response.agents.length,
          pageCount: response.internal.pages.length,
          toolCount: response.internal.tools.length
        });
        sendJson(res, 200, response);
        return;
      }
      if (payload.type === "manifest") {
        debugLog2(debug, "invalid manifest payload");
        sendJson(res, 400, {
          error: "Invalid manifest request",
          code: "INVALID_MANIFEST_REQUEST"
        });
        return;
      }
      debugLog2(debug, "unknown request type", { type: payload.type });
      sendJson(res, 400, { error: "Unknown request type" });
    } catch (error) {
      debugLog2(debug, "handler error", {
        error: error instanceof Error ? error.message : "Unknown error"
      });
      const parsed = asKonsierError(error);
      sendJson(res, parsed.statusCode, {
        error: parsed.message,
        code: parsed.code
      });
      if (next) {
        next(parsed);
      }
    }
  };
}
function debugLog2(debug, message, meta) {
  if (!debug || process.env.NODE_ENV !== "development") {
    return;
  }
  console.log("[konsier] handler", meta ? { message, ...meta } : { message });
}
function asManifestRequest(payload) {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    return null;
  }
  const obj = payload;
  if (obj.type !== "manifest") {
    return null;
  }
  const account = parseManifestAccount(obj.account);
  if (obj.account !== null && obj.account !== void 0 && account === null) {
    return null;
  }
  return {
    type: "manifest",
    account
  };
}
function parseManifestAccount(value) {
  if (value === null || value === void 0) {
    return null;
  }
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  const obj = value;
  if (!(typeof obj.id === "string" || typeof obj.id === "number") || typeof obj.name !== "string" || !obj.metadata || typeof obj.metadata !== "object" || Array.isArray(obj.metadata)) {
    return null;
  }
  return {
    id: obj.id,
    name: obj.name,
    metadata: obj.metadata
  };
}
function normalizeManifestAccount(account) {
  if (!account) {
    return null;
  }
  return {
    id: String(account.id),
    name: account.name,
    metadata: account.metadata
  };
}
function extractRawBody(req, rawBodyProperty) {
  const raw = req[rawBodyProperty] ?? req.body;
  if (Buffer.isBuffer(raw)) {
    return raw.toString("utf8");
  }
  if (typeof raw === "string") {
    return raw;
  }
  if (raw && typeof raw === "object") {
    return JSON.stringify(raw);
  }
  return "";
}
function parseBody(body, rawBody) {
  if (body && typeof body === "object" && !Array.isArray(body)) {
    return body;
  }
  if (typeof body === "string") {
    try {
      const parsed = JSON.parse(body);
      if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
        return parsed;
      }
      return null;
    } catch {
      return null;
    }
  }
  if (!rawBody) {
    return null;
  }
  try {
    const parsed = JSON.parse(rawBody);
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
      return parsed;
    }
    return null;
  } catch {
    return null;
  }
}
function sendJson(res, statusCode, body) {
  if (typeof res.status === "function") {
    res.status(statusCode);
  } else {
    res.statusCode = statusCode;
  }
  if (typeof res.json === "function") {
    res.json(body);
    return;
  }
  if (typeof res.setHeader === "function") {
    res.setHeader("content-type", "application/json");
  }
  const payload = JSON.stringify(body);
  if (typeof res.send === "function") {
    res.send(payload);
    return;
  }
  if (typeof res.end === "function") {
    res.end(payload);
  }
}

// src/page/verify.ts
function resolvePageRequest(options, input) {
  const debug = Boolean(options.debug);
  const pageUrl = parseRequestUrl(input.url);
  if (!pageUrl) {
    debugLog3(debug, "page auth failed", { reason: "INVALID_URL" });
    return unauthorizedResponse();
  }
  const launchToken = pageUrl.searchParams.get(PAGE_LAUNCH_QUERY_PARAM)?.trim() ?? "";
  if (launchToken) {
    const verifiedLaunch = verifyPageToken({
      apiKey: options.apiKey,
      token: launchToken
    });
    if (!verifiedLaunch.ok || verifiedLaunch.payload.type !== "launch") {
      debugLog3(debug, "page launch token rejected", {
        reason: verifiedLaunch.ok ? "INVALID_LAUNCH_TYPE" : verifiedLaunch.reason,
        pagePath: pageUrl.pathname
      });
      return unauthorizedResponse();
    }
    if (verifiedLaunch.payload.pagePath !== pageUrl.pathname) {
      debugLog3(debug, "page launch token path mismatch", {
        expectedPath: verifiedLaunch.payload.pagePath,
        requestPath: pageUrl.pathname
      });
      return unauthorizedResponse();
    }
    const sessionToken2 = createPageSessionToken({
      apiKey: options.apiKey,
      projectId: verifiedLaunch.payload.projectId,
      account: verifiedLaunch.payload.account,
      user: verifiedLaunch.payload.user,
      theme: verifiedLaunch.payload.theme,
      exp: Date.now() + (options.pageSessionTtlMs ?? DEFAULT_PAGE_SESSION_TTL_MS)
    });
    const cleanUrl = new URL(pageUrl.toString());
    cleanUrl.searchParams.delete(PAGE_LAUNCH_QUERY_PARAM);
    debugLog3(debug, "page launch token accepted", {
      pagePath: pageUrl.pathname,
      projectId: verifiedLaunch.payload.projectId,
      accountId: verifiedLaunch.payload.account?.id ?? null,
      userId: verifiedLaunch.payload.user.id ?? null
    });
    return {
      type: "response",
      status: 302,
      headers: {
        "cache-control": "no-store",
        location: `${cleanUrl.pathname}${cleanUrl.search}${cleanUrl.hash}`,
        "set-cookie": serializePageSessionCookie(
          sessionToken2,
          pageUrl.protocol === "https:",
          options.pageSessionTtlMs ?? DEFAULT_PAGE_SESSION_TTL_MS
        )
      }
    };
  }
  const sessionToken = readPageSessionCookie(input.headers);
  if (!sessionToken) {
    debugLog3(debug, "page session cookie missing", { pagePath: pageUrl.pathname });
    return unauthorizedResponse();
  }
  const verifiedSession = verifyPageToken({
    apiKey: options.apiKey,
    token: sessionToken
  });
  if (!verifiedSession.ok || verifiedSession.payload.type !== "session") {
    debugLog3(debug, "page session rejected", {
      reason: verifiedSession.ok ? "INVALID_SESSION_TYPE" : verifiedSession.reason,
      pagePath: pageUrl.pathname
    });
    return unauthorizedResponse();
  }
  const context = createPageContext(
    pageUrl.pathname,
    verifiedSession.payload.projectId,
    verifiedSession.payload.account,
    verifiedSession.payload.theme,
    verifiedSession.payload.user
  );
  debugLog3(debug, "page session resolved", {
    pagePath: context.pagePath,
    projectId: context.projectId,
    accountId: context.account?.id ?? null,
    userId: context.user.id ?? null
  });
  return {
    type: "authorized",
    context
  };
}
function unauthorizedResponse() {
  return {
    type: "response",
    status: 401,
    headers: {
      "cache-control": "no-store",
      "content-type": "text/plain; charset=utf-8"
    },
    body: "Unauthorized"
  };
}
function createPageContext(pagePath, projectId, account, theme, user) {
  const contextUser = {};
  if (user.id) {
    contextUser.id = user.id;
  }
  if (user.email) {
    contextUser.email = user.email;
  }
  if (user.name) {
    contextUser.name = user.name;
  }
  return {
    pagePath,
    projectId,
    account: normalizeAccount3(account),
    theme,
    user: contextUser
  };
}
function normalizeAccount3(account) {
  if (!account?.id || !account.name) {
    return null;
  }
  return {
    id: account.id,
    name: account.name,
    metadata: account.metadata ?? {}
  };
}
function parseRequestUrl(raw) {
  try {
    return new URL(raw);
  } catch {
    return null;
  }
}
function readPageSessionCookie(headers) {
  const cookieHeader = getHeaderValue(headers, "cookie");
  if (!cookieHeader) {
    return null;
  }
  const cookies = cookieHeader.split(";");
  for (const cookie of cookies) {
    const [rawName, ...rawValue] = cookie.trim().split("=");
    if (rawName === PAGE_SESSION_COOKIE_NAME) {
      const value = rawValue.join("=").trim();
      return value || null;
    }
  }
  return null;
}
function serializePageSessionCookie(value, secure, ttlMs) {
  const parts = [
    `${PAGE_SESSION_COOKIE_NAME}=${value}`,
    "Path=/",
    "HttpOnly",
    "SameSite=Lax",
    `Max-Age=${Math.max(1, Math.floor(ttlMs / 1e3))}`
  ];
  if (secure) {
    parts.push("Secure");
  }
  return parts.join("; ");
}
function debugLog3(debug, message, meta) {
  if (!debug || process.env.NODE_ENV !== "development") {
    return;
  }
  console.log("[konsier] pageAuth", meta ? { message, ...meta } : { message });
}

// src/tool.ts
import { toJSONSchema } from "zod";
var TOOL_NAME_REGEX = /^[a-zA-Z0-9_-]+$/;
function createTool(definition) {
  const name = definition.name?.trim();
  if (!name) {
    throw new KonsierError({
      code: "INVALID_TOOL_NAME",
      message: "Tool name is required.",
      statusCode: 400
    });
  }
  if (!TOOL_NAME_REGEX.test(name)) {
    throw new KonsierError({
      code: "INVALID_TOOL_NAME",
      message: `Tool name "${name}" contains invalid characters.`,
      statusCode: 400
    });
  }
  const description = definition.description?.trim();
  if (!description) {
    throw new KonsierError({
      code: "INVALID_TOOL_DESCRIPTION",
      message: `Tool "${name}" must include a description.`,
      statusCode: 400
    });
  }
  if (typeof definition.handler !== "function") {
    throw new KonsierError({
      code: "INVALID_TOOL_HANDLER",
      message: `Tool "${name}" handler must be a function.`,
      statusCode: 400
    });
  }
  const inputSchema = deriveInputSchema(name, definition.input);
  return {
    name,
    description,
    inputSchema,
    parseInput(input) {
      return parseToolInput(name, definition.input, input);
    },
    handler: definition.handler
  };
}
function parseToolInput(toolName, schema, input) {
  if (schema && typeof schema.safeParse === "function") {
    const parsed = schema.safeParse(input);
    if (parsed.success) {
      return parsed.data;
    }
    throw new KonsierError({
      code: "INVALID_TOOL_INPUT",
      message: `Tool "${toolName}" received invalid input: ${formatParseError(
        parsed.error
      )}`,
      statusCode: 400,
      details: parsed.error
    });
  }
  if (schema && typeof schema.parse === "function") {
    try {
      return schema.parse(input);
    } catch (error) {
      throw new KonsierError({
        code: "INVALID_TOOL_INPUT",
        message: `Tool "${toolName}" received invalid input: ${toErrorMessage(
          error
        )}`,
        statusCode: 400,
        details: error
      });
    }
  }
  return input;
}
function deriveInputSchema(toolName, schema) {
  if (schema && typeof schema.toJSONSchema === "function") {
    return schema.toJSONSchema();
  }
  if (schema && typeof schema.toJsonSchema === "function") {
    return schema.toJsonSchema();
  }
  if (isZodSchema(schema)) {
    try {
      const converted = toJSONSchema(schema);
      if (converted && typeof converted === "object" && !Array.isArray(converted)) {
        return converted;
      }
    } catch {
    }
  }
  return {
    type: "object",
    additionalProperties: true
  };
}
function isZodSchema(schema) {
  if (!schema || typeof schema !== "object") {
    return false;
  }
  const maybe = schema;
  return typeof maybe.safeParse === "function" && "_zod" in maybe;
}
function formatParseError(error) {
  if (!error) {
    return "validation failed";
  }
  if (typeof error === "string") {
    return error;
  }
  if (error instanceof Error) {
    return error.message;
  }
  if (typeof error === "object" && error !== null && "issues" in error && Array.isArray(error.issues)) {
    return error.issues.map((issue) => {
      if (issue && typeof issue === "object" && "message" in issue && typeof issue.message === "string") {
        return issue.message;
      }
      return "validation issue";
    }).join(", ");
  }
  return "validation failed";
}

// src/client.ts
function shouldDebugLog2(debug) {
  return debug && process.env.NODE_ENV === "development";
}
function maskSecret(value) {
  const normalized = value.trim();
  if (!normalized) {
    return "<empty>";
  }
  if (normalized.length <= 8) {
    return "***";
  }
  return `${normalized.slice(0, 4)}...${normalized.slice(-4)}`;
}
function normalizeEndpointUrl(raw) {
  if (typeof raw === "undefined") {
    return null;
  }
  const normalized = raw.trim().replace(/\/+$/, "");
  if (!normalized) {
    throw new KonsierError({
      code: "INVALID_ENDPOINT_URL",
      message: "Konsier endpointUrl must be a valid http(s) URL.",
      statusCode: 400
    });
  }
  let parsed;
  try {
    parsed = new URL(normalized);
  } catch {
    throw new KonsierError({
      code: "INVALID_ENDPOINT_URL",
      message: "Konsier endpointUrl must be a valid http(s) URL.",
      statusCode: 400
    });
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw new KonsierError({
      code: "INVALID_ENDPOINT_URL",
      message: "Konsier endpointUrl must use http or https.",
      statusCode: 400
    });
  }
  if (parsed.search || parsed.hash) {
    throw new KonsierError({
      code: "INVALID_ENDPOINT_URL",
      message: "Konsier endpointUrl must not include query or hash segments.",
      statusCode: 400
    });
  }
  return parsed.toString().replace(/\/+$/, "");
}
function endpointPath(endpointUrl) {
  const pathname = new URL(endpointUrl).pathname.replace(/\/+$/, "");
  return pathname || "/";
}
var Konsier = class {
  static tool = createTool;
  users;
  accounts;
  connections;
  apiKey;
  agents;
  internal;
  cloudClient;
  allowedClockSkewMs;
  endpointUrl;
  debug;
  constructor(options) {
    if (!options.apiKey?.trim()) {
      throw new KonsierError({
        code: "INVALID_API_KEY",
        message: "Konsier requires a non-empty apiKey.",
        statusCode: 400
      });
    }
    if ((!options.agents || Object.keys(options.agents).length === 0) && !options.internal) {
      throw new KonsierError({
        code: "INVALID_SURFACES",
        message: "Konsier requires at least one agent or internal definition.",
        statusCode: 400
      });
    }
    this.apiKey = options.apiKey;
    this.agents = options.agents ?? {};
    this.internal = options.internal ?? null;
    this.allowedClockSkewMs = DEFAULT_ALLOWED_CLOCK_SKEW_MS;
    this.endpointUrl = normalizeEndpointUrl(options.endpointUrl);
    this.debug = Boolean(options.debug);
    const cloudBaseUrl = resolveCloudBaseUrl({ debug: this.debug });
    const cloudClientOptions = {
      apiKey: this.apiKey,
      baseUrl: cloudBaseUrl,
      debug: this.debug
    };
    this.cloudClient = new CloudApiClient(cloudClientOptions);
    this.users = {
      get: async (input) => {
        return getUser(this.cloudClient, input);
      },
      link: async (input) => {
        return linkUser(this.cloudClient, input);
      }
    };
    this.accounts = {
      list: async () => {
        return listAccounts(this.cloudClient);
      },
      get: async (input) => {
        return getAccount(this.cloudClient, input);
      },
      link: async (input) => {
        return linkAccount(this.cloudClient, input);
      }
    };
    this.connections = {
      start: async (input) => {
        return startConnection(this.cloudClient, input);
      },
      complete: async (input) => {
        return completeConnection(this.cloudClient, input);
      }
    };
    this.validateAgentRegistry();
    this.validateInternalRegistry();
    if (shouldDebugLog2(this.debug)) {
      console.log("[konsier] initialized", {
        endpointUrl: this.endpointUrl,
        cloudBaseUrl,
        cloudBaseUrlEnv: process.env[ENV_CLOUD_BASE_URL] ?? null,
        apiKey: maskSecret(this.apiKey),
        agentRefs: Object.keys(this.agents),
        internalPages: this.internal && typeof this.internal !== "function" ? (this.internal.pages ?? []).length : null,
        internalTools: this.internal && typeof this.internal !== "function" ? (this.internal.tools ?? []).length : null
      });
    }
  }
  pageRequest(input) {
    return resolvePageRequest(
      {
        apiKey: this.apiKey,
        debug: this.debug
      },
      input
    );
  }
  webhookPath() {
    if (!this.endpointUrl) {
      throw new KonsierError({
        code: "ENDPOINT_URL_REQUIRED",
        message: "Konsier endpointUrl is required for webhook adapters.",
        statusCode: 400
      });
    }
    return endpointPath(this.endpointUrl);
  }
  webhookHandler() {
    return createHandler({
      apiKey: this.apiKey,
      allowedClockSkewMs: this.allowedClockSkewMs,
      debug: this.debug,
      handleToolCall: {
        debug: this.debug,
        resolveAgentConfig: (agent, account) => this.resolveAgentConfig(agent, account),
        resolveInternalDefinition: (account) => this.resolveInternalEntry({ account }),
        sendMessage: (input) => this.sendMessage(input)
      },
      handleResolveAgent: {
        resolveAgentConfig: (agent, account) => this.resolveAgentConfig(agent, account)
      },
      handleManifest: {
        listManifest: (account) => this.listManifest({ account })
      }
    });
  }
  async sendMessage(input) {
    await sendMessage(this.cloudClient, input);
  }
  async sync() {
    const payload = this.endpointUrl === null ? {} : { endpoint_url: this.endpointUrl };
    if (shouldDebugLog2(this.debug)) {
      console.log("[konsier] sync requested", payload);
    }
    await this.cloudClient.post("/agents/refresh", payload);
    if (shouldDebugLog2(this.debug)) {
      console.log("[konsier] sync succeeded", payload);
    }
  }
  async resolveAgentConfig(agentKey, account) {
    const resolved = await this.resolveAgentEntry(agentKey, account);
    this.assertValidAgentConfig(resolved, agentKey, 500);
    return resolved;
  }
  async listManifest(context) {
    const agents = await Promise.all(
      Object.keys(this.agents).map(async (ref) => {
        const resolved = await this.resolveAgentEntry(ref, context.account);
        this.assertValidAgentConfig(resolved, ref, 500);
        return {
          ref,
          name: typeof resolved.name === "string" && resolved.name.trim() ? resolved.name.trim() : ref,
          description: typeof resolved.description === "string" && resolved.description.trim() ? resolved.description.trim() : null
        };
      })
    );
    const internal = await this.resolveInternalEntry(context);
    return {
      agents,
      internal: {
        tools: this.manifestTools(internal.tools ?? [], "internal", 500),
        pages: this.pageDefinitions(internal.pages ?? [], "internal", 500)
      }
    };
  }
  async resolveAgentEntry(agentKey, account) {
    const entry = this.agents[agentKey];
    if (!entry) {
      throw new KonsierError({
        code: "AGENT_NOT_FOUND",
        message: `Agent "${agentKey}" is not registered in this SDK instance.`,
        statusCode: 404
      });
    }
    return typeof entry === "function" ? await entry({ account }) : entry;
  }
  async resolveInternalEntry(context) {
    if (!this.internal) {
      return {};
    }
    const resolved = typeof this.internal === "function" ? await this.internal(context) : this.internal;
    if (!resolved || typeof resolved !== "object" || Array.isArray(resolved)) {
      throw new KonsierError({
        code: "INVALID_INTERNAL_CONFIG",
        message: "internal must resolve to an object.",
        statusCode: 500
      });
    }
    return resolved;
  }
  validateAgentRegistry() {
    for (const [agentKey, entry] of Object.entries(this.agents)) {
      const key = agentKey.trim();
      if (!key) {
        throw new KonsierError({
          code: "INVALID_AGENT_KEY",
          message: "Agent keys must be non-empty strings.",
          statusCode: 400
        });
      }
      if (typeof entry === "function") {
        continue;
      }
      this.assertValidAgentConfig(entry, agentKey, 400);
    }
  }
  validateInternalRegistry() {
    if (!this.internal || typeof this.internal === "function") {
      return;
    }
    this.manifestTools(this.internal.tools ?? [], "internal", 400);
    this.pageDefinitions(this.internal.pages ?? [], "internal", 400);
  }
  assertValidAgentConfig(config, agentKey, statusCode) {
    if (!config?.systemPrompt?.trim()) {
      throw new KonsierError({
        code: "INVALID_AGENT_CONFIG",
        message: `Agent "${agentKey}" must include systemPrompt.`,
        statusCode
      });
    }
    this.manifestTools(config.tools, `agent "${agentKey}"`, statusCode);
  }
  manifestTools(tools, owner, statusCode) {
    if (!Array.isArray(tools)) {
      throw new KonsierError({
        code: "INVALID_TOOL",
        message: `${owner} tools must be an array.`,
        statusCode
      });
    }
    const names = /* @__PURE__ */ new Set();
    return tools.map((tool) => {
      this.validateTool(tool, owner, statusCode);
      if (names.has(tool.name)) {
        throw new KonsierError({
          code: "DUPLICATE_TOOL_NAME",
          message: `${owner} contains duplicate tool "${tool.name}".`,
          statusCode
        });
      }
      names.add(tool.name);
      return {
        name: tool.name,
        description: tool.description,
        input: tool.inputSchema
      };
    });
  }
  pageDefinitions(pages, owner, statusCode) {
    if (!Array.isArray(pages)) {
      throw new KonsierError({
        code: "INVALID_PAGE",
        message: `${owner} pages must be an array.`,
        statusCode
      });
    }
    const names = /* @__PURE__ */ new Set();
    return pages.map((page) => {
      const name = typeof page?.name === "string" && page.name.trim() ? page.name.trim() : "";
      const path = typeof page?.path === "string" && page.path.trim() ? page.path.trim() : "";
      if (!name || !path) {
        throw new KonsierError({
          code: "INVALID_PAGE",
          message: `${owner} pages must include non-empty name and path values.`,
          statusCode
        });
      }
      if (names.has(name)) {
        throw new KonsierError({
          code: "DUPLICATE_PAGE_NAME",
          message: `${owner} contains duplicate page "${name}".`,
          statusCode
        });
      }
      names.add(name);
      return { name, path };
    });
  }
  validateTool(tool, owner, statusCode) {
    if (!tool || typeof tool !== "object") {
      throw new KonsierError({
        code: "INVALID_TOOL",
        message: `${owner} contains an invalid tool entry.`,
        statusCode
      });
    }
    if (!tool.name || typeof tool.name !== "string") {
      throw new KonsierError({
        code: "INVALID_TOOL",
        message: `${owner} includes a tool with no name.`,
        statusCode
      });
    }
    if (typeof tool.handler !== "function" || typeof tool.parseInput !== "function") {
      throw new KonsierError({
        code: "INVALID_TOOL",
        message: `Tool "${tool.name}" is not a valid Konsier tool. Use Konsier.tool({...}).`,
        statusCode
      });
    }
  }
};
export {
  Konsier
};
